### PR TITLE
Digiline Filter-Injector tweaks

### PR DESCRIPTION
**Pull request by electrodude**
* Made Digiline Filter-Injector not pull a whole stack if the count is exactly 1
* Made Digiline Filter-Injector pull a whole stack if no count is specified
  * `default:dirt` still has a count of 1, but `{name="name"}` has no count

The first change is to not cause confusing behavior from the fact that sending a Digiline Filter-Injector `default:dirt 2` will cause it to output a stack of 2 dirt blocks, while sending it `default:dirt 1` or `default:dirt` would cause it to output **a whole stack** and not just 1 dirt block.  This changes it to only output only 1 dirt block.

The second change is to add a way to get a Digiline Filter-Injector to output a whole stack, since the first change removed this ability.

UPDATE: Also, added metadata, wear, and group matching.  See [comment below](https://github.com/minetest-mods/pipeworks/pull/141#issuecomment-245760457)

### CODE REVIEW

**Comment by sofar** [comment]
Looks OK to me.

**Comment by electrodude** [comment]
Don't merge this yet; I want to make it able to match wear and metadata if specified.

**Comment by sofar** [comment]
Any updates to this?

File: filter-injector.lua [hunk 1]
```diff
@@ -206,13 +207,13 @@ local function punch_filter(data, filtpos, filtnode, msg)
 			end
 
 			if type(msg.name) == "string" then
-				table.insert(filters, {name = msg.name, count = tonumber(msg.count) or 1})
+				table.insert(filters, {name = msg.name, count = tonumber(msg.count)})
```
Comment (HybridDog) [inline filter-injector.lua]: filters[#filters+1] = {name = msg.name, count = tonumber(msg.count)}

**Comment by Ekdohibs** [comment]
Looks ok to me too; this code would need some serious refactoring later: ```grabAndFire``` has 14 arguments!

**Comment by electrodude** [comment]
Sorry I never got around to testing the wear and metadata matching feature that I was going to add to this PR.

Please merge this PR as it is now.  It fixes all of the serious problems that I'm aware of with my original implementation of the Digiline Filter-Injector.  I'll eventually make a new PR to add the more precise matching feature I have implemented but still haven't gotten around to testing yet.

`grabAndFire`'s argument list is quite a sight.  It's only ever called by `punch_filter`, so the best way to fix it would probably be to make it a closure inside of `punch_filter` right before the loop in which it's called, so it can share all of `punch_filter`'s locals without passing them.  Or, it could instead be made to get as much as possible out of meta instead of having it passed to it, but that would be somewhat slower.  Maybe I could refactor it (and do the `filters[#filters+1]` thing) in a few weeks or so, when I get around to testing wear and metadata matching.

**Comment by electrodude** [comment]
I just rebased this on to eee40e7679cfc759dbe2122151d69b4e3a53f197.  I'll finish testing wear and metadata matching now and then push that to this PR.

**Comment by electrodude** [comment]
So, I finally got around to finishing and testing wear, metadata, and group matching.

You can now do things like match anything that's more than half worn out (`{wear={32768,nil}}`), grab 3 of anything in group `stick` (`{group="stick", count=3}`), grab an item with an exact metadata (`{name="mod:item", metadata="whatever"}`), etc.  Does this warrant an increase in crafting cost?

Can someone who actually understands rotation and random modes please make sure they still work properly?  I thought I broke them, but I now realize that they just work wierdly.  It would make more sense to me if random mode meant take the first possible stack matching a random filter, rather than a random stack matching the first possible filter.  Maybe there should be two buttons?

I tested this relatively thoroughly, but more testing is welcome.  Please merge this when you're satisfied it works.

**Comment by electrodude** [comment]
This has been done for more than a month and has still not been merged yet.  Do I need to do anything else before this can be merged, or is everyone just busy with real life?

**Comment by sofar** [comment]
Ah, I didn't see the Sep 8 update to this. *looking*

**Comment by sofar** [comment]
I'm thinking this is fine as is. No idea on the random thing - we can work it out later if it proves to be a problem?

Squash this or retain the 2 commits separately (probably better?)

**Comment by electrodude** [comment]
Please leave the two commits separate.  One fixes stuff that should have been in #128 (add Digiline Filter-Injector) to begin with, and the other adds more features without changing any previous behavior.

Also, please close bug #138 as fixed when you merge this, since the first commit of this PR fixes that bug.

I'm pretty sure that the random thing was just that I misunderstood how it was intended to behave.  I still think the way it works is counterintuitive and I couldn't see how it might be useful last time I looked, so it might be a good idea to add another button to the GUI of all three Filter-Injectors to let you change the filter preference order as well as the inventory preference order, but that probably belongs in another PR.

**Comment by electrodude** [comment]
Thanks!

**Pull request merged**